File: Task 9/main.js
import 'dotenv/config';
import OpenAI from 'openai';
import chalk from 'chalk';
import MarkdownIt from 'markdown-it';
import { promises as fs } from 'fs';
import readline from 'readline';

// Initialize OpenAI client
const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY
});

// Initialize markdown parser
const md = new MarkdownIt();

// Create readline interface
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Promisify readline question
const question = (query) => new Promise((resolve) => rl.question(query, resolve));

async function generateReport(inputText, isServiceName) {
  try {
    console.log(chalk.blue('\nPreparing API request...'));
    
    // Prepare the prompt
    const prompt = isServiceName
      ? `Generate a comprehensive analysis report about ${inputText} service.`
      : `Generate a comprehensive analysis report about the following service:\n\n${inputText}`;

    const fullPrompt = `${prompt}\n\nPlease include the following sections:
1. Brief History (founding year, milestones)
2. Target Audience (primary user segments)
3. Core Features (2-4 key functionalities)
4. Unique Selling Points (key differentiators)
5. Business Model (how the service makes money)
6. Tech Stack Insights (technologies used)
7. Perceived Strengths (standout features)
8. Perceived Weaknesses (limitations)

Format the response in Markdown.`;

    console.log(chalk.blue('Calling OpenAI API... This may take a minute...'));
    
    // Call OpenAI API
    const response = await openai.chat.completions.create({
      model: "gpt-4.1-mini",
      messages: [
        { role: "system", content: "You are a professional business and technical analyst." },
        { role: "user", content: fullPrompt }
      ]
    });

    console.log(chalk.green('Successfully received response from OpenAI!'));
    
    return response.choices[0].message.content;
  } catch (error) {
    console.error(chalk.red(`\nError generating report: ${error.message}`));
    console.error(chalk.yellow('Full error details:', error));
    return null;
  }
}

async function main() {
  try {
    console.log(chalk.blue.bold('Service Analysis Report Generator\n'));
    console.log('Choose input type:');
    console.log('1. Service name (e.g., "Spotify", "Notion")');
    console.log('2. Service description (text)\n');

    const choice = await question('Enter your choice (1 or 2): ');

    let inputText;
    let isServiceName;

    if (choice === '1') {
      inputText = await question('\nEnter service name: ');
      isServiceName = true;
    } else if (choice === '2') {
      console.log('\nEnter service description (press Enter twice to finish):');
      const lines = [];
      
      while (true) {
        const line = await question('');
        if (line === '') break;
        lines.push(line);
      }
      
      inputText = lines.join('\n');
      isServiceName = false;
    } else {
      console.log(chalk.red('\nInvalid choice. Please run the program again.'));
      process.exit(1);
    }

    console.log(chalk.green('\nStarting report generation...'));
    console.log(chalk.blue(`API Key status: ${process.env.OPENAI_API_KEY ? 'Found' : 'Not found'}`));
    
    const report = await generateReport(inputText, isServiceName);

    if (report) {
      // Display the report
      console.log('\nReport:');
      console.log(report);

      // Save the report
      const filename = `report_${inputText.split(' ')[0].toLowerCase()}.md`;
      await fs.writeFile(filename, report, 'utf8');
      console.log(chalk.green(`\nReport saved to ${filename}`));
    }
  } catch (error) {
    console.error(chalk.red(`\nError in main: ${error.message}`));
    console.error(chalk.yellow('Full error details:', error));
  } finally {
    rl.close();
  }
}

// Run the program
main(); 

File: Task 9/main.py
import os
from dotenv import load_dotenv
from openai import OpenAI
from rich.console import Console
from rich.markdown import Markdown

# Load environment variables
load_dotenv()

# Initialize OpenAI client
client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))

# Initialize Rich console
console = Console()

def generate_report(input_text: str, is_service_name: bool = False) -> str:
    """
    Generate a comprehensive report about a service using OpenAI API.
    
    Args:
        input_text (str): Service name or description
        is_service_name (bool): Whether the input is a service name
        
    Returns:
        str: Markdown formatted report
    """
    # Prepare the prompt
    if is_service_name:
        prompt = f"Generate a comprehensive analysis report about {input_text} service."
    else:
        prompt = f"Generate a comprehensive analysis report about the following service:\n\n{input_text}"
    
    prompt += """
    
    Please include the following sections:
    1. Brief History (founding year, milestones)
    2. Target Audience (primary user segments)
    3. Core Features (2-4 key functionalities)
    4. Unique Selling Points (key differentiators)
    5. Business Model (how the service makes money)
    6. Tech Stack Insights (technologies used)
    7. Perceived Strengths (standout features)
    8. Perceived Weaknesses (limitations)
    
    Format the response in Markdown."""

    try:
        # Call OpenAI API
        response = client.chat.completions.create(
            model="gpt-4.1-mini",
            messages=[
                {"role": "system", "content": "You are a professional business and technical analyst."},
                {"role": "user", "content": prompt}
            ]
        )
        
        return response.choices[0].message.content
    except Exception as e:
        return f"Error generating report: {str(e)}"

def main():
    """Main function to run the application."""
    console.print("[bold blue]Service Analysis Report Generator[/bold blue]")
    console.print("\nChoose input type:")
    console.print("1. Service name (e.g., 'Spotify', 'Notion')")
    console.print("2. Service description (text)")
    
    choice = input("\nEnter your choice (1 or 2): ").strip()
    
    if choice == "1":
        input_text = input("\nEnter service name: ").strip()
        is_service_name = True
    elif choice == "2":
        console.print("\nEnter service description (press Enter twice to finish):")
        lines = []
        while True:
            line = input()
            if line == "":
                break
            lines.append(line)
        input_text = "\n".join(lines)
        is_service_name = False
    else:
        console.print("[bold red]Invalid choice. Please run the program again.[/bold red]")
        return

    console.print("\n[bold green]Generating report...[/bold green]")
    report = generate_report(input_text, is_service_name)
    
    # Display the report
    console.print(Markdown(report))
    
    # Save the report
    filename = f"report_{input_text.split()[0].lower()}.md"
    with open(filename, "w", encoding="utf-8") as f:
        f.write(report)
    console.print(f"\n[bold green]Report saved to {filename}[/bold green]")

if __name__ == "__main__":
    main() 